tests/lenient-matching.test.ts

```diff
@@ -0,0 +1,69 @@
+import { describe, expect, it, vi } from "vitest";
+import { z } from "zod";
+
+import { loadConfig } from "@/index";
+
+describe("loadConfig()", () => {
+    it("should perform lenient matching", async () => {
+        const MyConfig = z.object({
+            foo: z.object({
+                enabled: z.boolean().default(true),
+                nestedProp: z.string(),
+                OTHER_PROP: z.string(),
+            }),
+        });
+
+        const mockedLogger = { warn: vi.fn() };
+        const config = await loadConfig({
+            logger: mockedLogger,
+            schema: MyConfig,
+            lenientMatching: true,
+            adapters: [
+                inlineAdapter({ FOO: { NESTED_PROP: "Foo!", otherProp: "Test" } }),
+            ],
+        });
+
+        expect(config).toEqual({
+            foo: {
+                enabled: true,
+                nestedProp: "Foo!",
+                OTHER_PROP: "Test",
+            },
+        });
+    });
+
+    it("lenient matching should preserve order of adapter", async () => {
+        const MyConfig = z.object({
+            foo: z.object({
+                enabled: z.boolean().default(true),
+                nestedProp: z.string(),
+                OTHER_PROP: z.string(),
+            }),
+        });
+
+        const mockedLogger = { warn: vi.fn() };
+        const config = await loadConfig({
+            logger: mockedLogger,
+            schema: MyConfig,
+            lenientMatching: true,
+            adapters: [
+                inlineAdapter({ FOO: { nestedProp: "Foo!", otherProp: "Test" } }),
+                inlineAdapter({ FOO: { NESTED_PROP: "Override1", otherProp: "Override1" } }),
+                inlineAdapter({ foo: { NESTED_PROP: "Override2", otherProp: "Override2" } }),
+            ],
+        });
+
+        expect(config).toEqual({
+            foo: {
+                enabled: true,
+                nestedProp: "Override2",
+                OTHER_PROP: "Override2",
+            },
+        });
+    });
+});
+
+// candidate for the library?
+function inlineAdapter(source: Record<string, unknown>) {
+    return { name: 'inline', read: () => source };
+}
```

Comment:
This function might be nice to add to the library. When chaining multiple adapters as functions, it looks a bit awkward to express an inline source adapter as object literal.

Reply:
Why not, yeah let's add an inlineAdapter, something like:
```

export function inlineAdapter(name: string, read: () => Promise<Record<string, unknown>>, options?: BaseAdapterProps): Adapter {
  return {
    name,
    read,
    ...options,
  };
}
```

then in the tests it will looks like: `inlineAdapter("inline", async () => ({ FOO: { NESTED_PROP: "Foo!", otherProp: "Test" } })),`. As soon sync configs are supported the async can be removed!

Reply:
I do have use a lot of cases where we need to import a raw value from a dynamic import (since we are using esbuild), and I was also writing a quick adapter that just takes a value coming from a custom function call. It would be nice to be able to define as a nice clean inline function.

This could then also serve as a low-level building bricks (high-order function) for a composable API.

I would then just call it `adapter()`:

```
export function adapter(read: () => Promise<Record<string, unknown>>, name: string?, options?: BaseAdapterProps): Adapter {
  return {
    read,
    name ?? "anonymous", // or make completely optional, only used in logger.
    ...options,
  };
}
```